src/search-manager/SearchManager.ts

```diff
@@ -36,19 +40,88 @@ export interface SearchContextData {
 
 const SearchContext = React.createContext<SearchContextData | undefined>(undefined);
 
+/**
+ * Hook which lets you store state variables in the URL search parameters.
+ *
+ * It wraps useState with functions that get/set a query string
+ * search parameter when returning/setting the state variable.
+ *
+ */
+function useStateWithUrlSearchParam<Type>(
+  defaultValue: Type,
+  paramName: string,
+  // Returns the Type equivalent of the given string value, or
+  // undefined if value is invalid.
+  fromString: (value: string | null) => Type | undefined,
+  // Returns the string equivalent of the given Type value.
+  // Returning empty string/undefined will clear the url search paramName.
+  toString: (value: Type) => string | undefined,
+): [value: Type, setter: React.Dispatch<React.SetStateAction<Type>>] {
+  const [searchParams, setSearchParams] = useSearchParams();
+  const [stateValue, stateSetter] = React.useState<Type>(defaultValue);
+
+  // The converted search parameter value takes precedence over the state value.
+  const returnValue: Type = fromString(searchParams.get(paramName)) ?? stateValue ?? defaultValue;
+
+  // Before updating the state value, update the url search parameter
+  const returnSetter: React.Dispatch<React.SetStateAction<Type>> = ((value: Type) => {
+    const paramValue: string = toString(value) ?? '';
+    if (paramValue) {
+      searchParams.set(paramName, paramValue);
+    } else {
+      // If no paramValue, remove it from the search params, so
+      // we don't get dangling parameter values like ?paramName=
+      // Another way to decide this would be to check value === defaultValue,
+      // and ensure that default values are never stored in the search string.
+      searchParams.delete(paramName);
+    }
+    setSearchParams(searchParams, { replace: true });
+    return stateSetter(value);
+  });
```

Comment:
OK, a few minor things here:
1. It's always a red flag to me when I see two different state variables that need to be kept in sync (`searchParams` and `stateValue` here). Do we actually need `stateValue` at all? It seems to me that `searchParams` _is_ the state and in my tests it works totally fine without `stateValue`.
2. React hooks usually [guarantee](https://legacy.reactjs.org/docs/hooks-faq.html#what-can-i-do-if-my-effect-dependencies-change-too-often) that "setter" functions like `setSearchParams` and `stateSetter` have a stable identity (never change), so it's good to follow this practice by making our `returnSetter` function here so it never changes. You can do that by wrapping it in `useCallback` and getting the "old" params from `setSearchParams` by passing it a function, rather than referencing the `searchParams` variable. This is also how react-router [makes `setSearchParams` itself stable](https://github.com/remix-run/react-router/blob/0f2b167e9b862d5e6b3425438787c8e7b39197f4/packages/react-router-dom/index.tsx#L1496).
3. It's not good practice to directly mutate a state object like `searchParams`, e.g. by calling `searchParams.set(...)`. That could lead to various bugs that are hard to track down. Instead, clone it using `const newSearchParams = new URLSearchParams(prevParams);` and mutate that.

```suggestion
  // The converted search parameter value takes precedence over the state value.
  const returnValue: Type = fromString(searchParams.get(paramName)) ?? defaultValue;

  // Function to update the url search parameter
  const returnSetter: React.Dispatch<React.SetStateAction<Type>> = React.useCallback((value: Type) => {
    setSearchParams((prevParams) => {
      const paramValue: string = toString(value) ?? '';
      const newSearchParams = new URLSearchParams(prevParams);
      if (paramValue) {
        newSearchParams.set(paramName, paramValue);
      } else {
        // If no paramValue, remove it from the search params, so
        // we don't get dangling parameter values like ?paramName=
        // Another way to decide this would be to check value === defaultValue,
        // and ensure that default values are never stored in the search string.
        newSearchParams.delete(paramName);
      }
      return newSearchParams;
    }, { replace: true });
  }, [setSearchParams]);
```

Reply:
Thank you for this information @bradenmacdonald ! I still have a lot to learn about React..